// @progress/kendo-angular-grid/dist/npm/column.component.js
"use strict";
var core_1 = require('@angular/core');
var cell_template_directive_1 = require('./cell-template.directive');
var header_template_directive_1 = require('./header-template.directive');
var footer_template_directive_1 = require('./footer-template.directive');
/**
 * Represents the columns of the Kendo UI Grid component for Angular.
 *
 * @example
 * ```ts-preview
 *
 * @@Component({
 *    selector: 'my-app',
 *    template: `
 *        <kendo-grid [data]="gridData">
 *          <kendo-grid-column field="ProductID" title="Product ID" width="120">
 *          </kendo-grid-column>
 *          <kendo-grid-column field="ProductName" title="Product Name">
 *          </kendo-grid-column>
 *          <kendo-grid-column field="UnitPrice" title="Unit Price" width="230">
 *          </kendo-grid-column>
 *          <kendo-grid-column field="Discontinued" width="120">
 *              <template kendoCellTemplate let-dataItem>
 *                  <input type="checkbox" [checked]="dataItem.Discontinued" disabled/>
 *              </template>
 *          </kendo-grid-column>
 *        </kendo-grid>
 *    `
 * })
 *
 * class AppComponent {
 *    private gridData: any[];
 *
 *    constructor() {
 *        this.gridData = products;
 *    }
 * }
 *
 * const products = [{
 *    "ProductID": 1,
 *    "ProductName": "Chai",
 *    "UnitPrice": 18.0000,
 *    "Discontinued": true
 *  }, {
 *    "ProductID": 2,
 *    "ProductName": "Chang",
 *    "UnitPrice": 19.0000,
 *    "Discontinued": false
 *  }
 * ];
 *
 * ```
 */
var ColumnComponent = (function () {
    function ColumnComponent() {
        /**
         * Allows the column headers to be clicked and the `sortChange` event emitted.
         * You have to handle the `sortChange` event yourself and sort the data.
         */
        this.sortable = true;
    }
    Object.defineProperty(ColumnComponent.prototype, "width", {
        get: function () { return this._width; },
        /**
         * The width of the column in pixels.
         */
        set: function (value) {
            this._width = parseInt(value, 10);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ColumnComponent.prototype, "templateRef", {
        get: function () {
            return this.template ? this.template.templateRef : undefined;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ColumnComponent.prototype, "headerTemplateRef", {
        get: function () {
            return this.headerTemplate ? this.headerTemplate.templateRef : undefined;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ColumnComponent.prototype, "footerTemplateRef", {
        get: function () {
            return this.footerTemplate ? this.footerTemplate.templateRef : undefined;
        },
        enumerable: true,
        configurable: true
    });
    ColumnComponent.decorators = [
        { type: core_1.Component, args: [{
                    selector: 'kendo-grid-column',
                    template: ""
                },] },
    ];
    /** @nocollapse */
    ColumnComponent.ctorParameters = [];
    ColumnComponent.propDecorators = {
        'field': [{ type: core_1.Input },],
        'title': [{ type: core_1.Input },],
        'format': [{ type: core_1.Input },],
        'width': [{ type: core_1.Input },],
        'locked': [{ type: core_1.Input },],
        'sortable': [{ type: core_1.Input },],
        'style': [{ type: core_1.Input },],
        'headerStyle': [{ type: core_1.Input },],
        'footerStyle': [{ type: core_1.Input },],
        'cssClass': [{ type: core_1.Input, args: ['class',] },],
        'headerClass': [{ type: core_1.Input },],
        'footerClass': [{ type: core_1.Input },],
        'template': [{ type: core_1.ContentChild, args: [cell_template_directive_1.CellTemplateDirective,] },],
        'headerTemplate': [{ type: core_1.ContentChild, args: [header_template_directive_1.HeaderTemplateDirective,] },],
        'footerTemplate': [{ type: core_1.ContentChild, args: [footer_template_directive_1.FooterTemplateDirective,] },],
    };
    return ColumnComponent;
}());
exports.ColumnComponent = ColumnComponent;
