// @progress/kendo-angular-grid/dist/npm/grid.module.js
"use strict";
var core_1 = require('@angular/core');
var common_1 = require('@angular/common');
var grid_component_1 = require('./grid.component');
var column_component_1 = require('./column.component');
var toolbar_component_1 = require('./toolbar.component');
var cell_template_directive_1 = require('./cell-template.directive');
var header_template_directive_1 = require('./header-template.directive');
var footer_template_directive_1 = require('./footer-template.directive');
var detail_template_directive_1 = require('./detail-template.directive');
var list_component_1 = require('./list.component');
var pager_component_1 = require('./pager.component');
var shared_module_1 = require('./shared.module');
var kendo_angular_intl_1 = require('@progress/kendo-angular-intl');
var exportedModules = [
    grid_component_1.GridComponent,
    column_component_1.ColumnComponent,
    toolbar_component_1.ToolbarComponent,
    cell_template_directive_1.CellTemplateDirective,
    header_template_directive_1.HeaderTemplateDirective,
    footer_template_directive_1.FooterTemplateDirective,
    detail_template_directive_1.DetailTemplateDirective
];
var declarations = [
    grid_component_1.GridComponent,
    toolbar_component_1.ToolbarComponent,
    list_component_1.ListComponent,
    pager_component_1.PagerComponent
];
/**
 * Represents the [NgModule](https://angular.io/docs/ts/latest/guide/ngmodule.html)
 * definition for the Grid component.
 *
 * @example
 *
 * ```ts-no-run
 * // Import the Grid module
 * import { GridModule } from '@progress/kendo-angular-grid';
 *
 * // The browser platform with a compiler
 * import { platformBrowserDynamic } from '@angular/platform-browser-dynamic';
 *
 * import { NgModule } from '@angular/core';
 *
 * // Import the app component
 * import { AppComponent } from './app.component';
 *
 * // Define the app module
 * @@NgModule({
 *     declarations: [AppComponent], // declare app component
 *     imports:      [BrowserModule, GridModule], // import Grid module
 *     bootstrap:    [AppComponent]
 * })
 * export class AppModule {}
 *
 * // Compile and launch the module
 * platformBrowserDynamic().bootstrapModule(AppModule);
 *
 * ```
 */
var GridModule = (function () {
    function GridModule() {
    }
    GridModule.decorators = [
        { type: core_1.NgModule, args: [{
                    declarations: [declarations],
                    exports: [exportedModules],
                    imports: [common_1.CommonModule, shared_module_1.SharedModule],
                    providers: [
                        { provide: kendo_angular_intl_1.IntlService, useClass: kendo_angular_intl_1.CldrIntlService }
                    ]
                },] },
    ];
    /** @nocollapse */
    GridModule.ctorParameters = [];
    return GridModule;
}());
exports.GridModule = GridModule;
