// @progress/kendo-angular-grid/dist/es/list.component.js
import { Component, Input, Output, EventEmitter, ViewChild, Inject, OpaqueToken } from '@angular/core';
import { Observable } from 'rxjs/Observable';
import { Subject } from 'rxjs/Subject';
import 'rxjs/add/observable/fromEvent';
import 'rxjs/add/operator/map';
import 'rxjs/add/operator/merge';
import 'rxjs/add/operator/filter';
import 'rxjs/add/operator/do';
import 'rxjs/add/operator/distinctUntilChanged';
import 'rxjs/add/operator/debounceTime';
import { RowHeightService } from './row-height.service';
import { ScrollerService, PageAction, ScrollAction } from './scroller.service';
import { isChanged } from './utils';
import { DetailsService } from './details.service';
import { ColumnsContainer } from './columns-container';
/**
 * @hidden
 */
export var SCROLLER_FACTORY_TOKEN = new OpaqueToken('grid-scroll-service-factory');
/**
 * @hidden
 */
export function DEFAULT_SCROLLER_FACTORY(observable) {
    return new ScrollerService(observable);
}
;
/**
 * @hidden
 */
var wheelDeltaY = function (e) {
    var deltaY = e.wheelDeltaY;
    if (e.wheelDelta && (deltaY === undefined || deltaY)) {
        return e.wheelDelta;
    }
    else if (e.detail && e.axis === e.VERTICAL_AXIS) {
        return (-e.detail) * 10;
    }
    return 0;
};
/**
 * @hidden
 */
export var ListComponent = (function () {
    function ListComponent(scrollerFactory, detailsService) {
        var _this = this;
        this.skip = 0;
        this.columns = new ColumnsContainer(function () { return []; });
        this.pageChange = new EventEmitter();
        this.containerScroll = new EventEmitter();
        this.dispatcher = new Subject();
        this.PAGE_EVENT_TIMEOUT = 200;
        this.scroller = scrollerFactory(this.dispatcher);
        this.detailsSubscription = detailsService.changes.subscribe(function (x) { return _this.detailExpand(x); });
    }
    Object.defineProperty(ListComponent.prototype, "lockedColumns", {
        get: function () {
            return this.columns.lockedColumns;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ListComponent.prototype, "nonLockedColumns", {
        get: function () {
            return this.columns.nonLockedColumns;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ListComponent.prototype, "lockedWidth", {
        get: function () {
            return this.lockedColumns.reduce(function (prev, curr) { return prev + (curr.width || 0); }, 0);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ListComponent.prototype, "nonLockedWidth", {
        get: function () {
            if (this.lockedColumns.length) {
                return this.nonLockedColumns.reduce(function (prev, curr) { return prev + (curr.width || 0); }, 0);
            }
            return undefined;
        },
        enumerable: true,
        configurable: true
    });
    ListComponent.prototype.ngOnInit = function () {
        this.init();
    };
    ListComponent.prototype.ngOnChanges = function (changes) {
        if (isChanged("total", changes)) {
            this.init();
        }
    };
    ListComponent.prototype.ngAfterViewInit = function () {
        var _this = this;
        this.container.nativeElement.scrollTop = this.rowHeightService.offset(this.skip);
        this.containerScrollSubscription = Observable
            .fromEvent(this.container.nativeElement, 'scroll')
            .map(function (event) { return event.target; })
            .do(this.onContainerScroll.bind(this))
            .subscribe(this.dispatcher);
        if (this.lockedContainer) {
            this.containerScrollSubscription = this.containerScrollSubscription
                .add(Observable.fromEvent(this.lockedContainer.nativeElement, 'mousewheel')
                .merge(Observable.fromEvent(this.lockedContainer.nativeElement, 'DOMMouseScroll'))
                .filter(function (event) { return !event.ctrlKey; })
                .map(function (event) { return wheelDeltaY(event); })
                .distinctUntilChanged()
                .subscribe(function (x) { return _this.container.nativeElement.scrollTop -= x; }));
        }
    };
    ListComponent.prototype.ngOnDestroy = function () {
        if (this.scroller) {
            this.scrollSubscription.unsubscribe();
            this.scroller.destroy();
        }
        if (this.containerScrollSubscription) {
            this.containerScrollSubscription.unsubscribe();
        }
        if (this.resizeSubscription) {
            this.resizeSubscription.unsubscribe();
        }
        if (this.detailsSubscription) {
            this.detailsSubscription.unsubscribe();
        }
    };
    ListComponent.prototype.init = function () {
        this.rowHeightService = new RowHeightService(this.total, this.rowHeight, this.detailRowHeight);
        this.totalHeight = this.rowHeightService.totalHeight();
        var observable = this.scroller
            .create(this.rowHeightService, this.skip, this.take, this.total);
        this.scrollSubscription = observable
            .filter(function (x) { return x instanceof PageAction; })
            .debounceTime(this.PAGE_EVENT_TIMEOUT)
            .subscribe(this.pageChange)
            .add(observable
            .filter(function (x) { return x instanceof ScrollAction; })
            .subscribe(this.scroll.bind(this)));
    };
    ListComponent.prototype.detailExpand = function (_a) {
        var index = _a.index, expand = _a.expand;
        if (expand) {
            this.rowHeightService.expandDetail(index);
        }
        else {
            this.rowHeightService.collapseDetail(index);
        }
        this.totalHeight = this.rowHeightService.totalHeight();
    };
    ListComponent.prototype.scroll = function (_a) {
        var offset = _a.offset;
        this.style = { transform: "translateY(" + offset + "px)" };
        //this.style = { top: offset + "px" };
    };
    ListComponent.prototype.onContainerScroll = function (_a) {
        var scrollTop = _a.scrollTop, scrollLeft = _a.scrollLeft;
        if (this.lockedContainer) {
            this.lockedContainer.nativeElement.scrollTop = scrollTop;
        }
        this.containerScroll.emit({ scrollTop: scrollTop, scrollLeft: scrollLeft });
    };
    ListComponent.decorators = [
        { type: Component, args: [{
                    providers: [
                        {
                            provide: SCROLLER_FACTORY_TOKEN,
                            useValue: DEFAULT_SCROLLER_FACTORY
                        }],
                    selector: 'kendo-grid-list',
                    template: "\n    <div #lockedContainer class=\"k-grid-content-locked\" [style.height.px]=\"height\"\n        *ngIf=\"lockedColumns.length\" [style.width.px]=\"lockedWidth\">\n        <table [ngStyle]=\"style\">\n            <colgroup kendoGridColGroup\n                [columns]=\"lockedColumns\"\n                [detailTemplate]=\"detailTemplate\">\n            </colgroup>\n            <tbody kendoGridTableBody\n                [data]=\"data\"\n                [columns]=\"lockedColumns\"\n                [detailTemplate]=\"detailTemplate\"\n                [skip]=\"skip\"\n                [selectable]=\"selectable\">\n            </tbody>\n        </table>\n        <div class=\"k-height-container\">\n            <div [style.height.px]=\"totalHeight\"></div>\n        </div>\n    </div><div #container class=\"k-grid-content k-virtual-content\"\n        [style.height.px]=\"height\"\n        [kendoGridResizableContainer]=\"lockedColumns.length\"\n        [lockedWidth]=\"lockedWidth + 1\">\n        <table [ngStyle]=\"style\" [style.width.px]=\"nonLockedWidth\">\n            <colgroup kendoGridColGroup\n                [columns]=\"nonLockedColumns\"\n                [detailTemplate]=\"detailTemplate\">\n            </colgroup>\n            <tbody kendoGridTableBody\n                [data]=\"data\"\n                [columns]=\"nonLockedColumns\"\n                [detailTemplate]=\"detailTemplate\"\n                [skip]=\"skip\"\n                [selectable]=\"selectable\">\n            </tbody>\n        </table>\n        <div class=\"k-height-container\">\n            <div [style.height.px]=\"totalHeight\"></div>\n        </div>\n    </div>"
                },] },
    ];
    /** @nocollapse */
    ListComponent.ctorParameters = [
        { type: undefined, decorators: [{ type: Inject, args: [SCROLLER_FACTORY_TOKEN,] },] },
        { type: DetailsService, },
    ];
    ListComponent.propDecorators = {
        'data': [{ type: Input },],
        'total': [{ type: Input },],
        'height': [{ type: Input },],
        'rowHeight': [{ type: Input },],
        'detailRowHeight': [{ type: Input },],
        'take': [{ type: Input },],
        'skip': [{ type: Input },],
        'columns': [{ type: Input },],
        'detailTemplate': [{ type: Input },],
        'selectable': [{ type: Input },],
        'pageChange': [{ type: Output },],
        'containerScroll': [{ type: Output },],
        'container': [{ type: ViewChild, args: ["container",] },],
        'lockedContainer': [{ type: ViewChild, args: ["lockedContainer",] },],
    };
    return ListComponent;
}());
